File: web/prisoner.py
"""Iterated Prisoner's Dilemma — web-native 2-player game engine, WebSocket handler, API router.

Mirrors the RPS/tic-tac-toe WS-game pattern (in-memory rooms, WebPlayer dataclass,
_send/_broadcast helpers, a WS handler, and a stale-room cleanup coroutine), but
rooms are created straight from the browser — no Discord token needed. This is a
free game (no coins).

Like RPS this is SIMULTANEOUS, not turn-based: each round both players secretly
pick Cooperate or Defect at the same time. A player's move is kept private until
*both* are in, then the round scores with the classic payoff matrix and each
player's running total ticks up. It's *iterated* — 10 rounds — and after round 10
the higher total wins (ties are possible).

Payoff (row = you, col = opponent):
  both Cooperate → +3 / +3
  both Defect    → +1 / +1
  you Cooperate, they Defect → +0 (you) / +5 (them)
  you Defect, they Cooperate → +5 (you) / +0 (them)
"""

import asyncio
import logging
import random
import secrets
import string
import time
from dataclasses import dataclass, field

from fastapi import APIRouter, HTTPException, WebSocket, WebSocketDisconnect
from pydantic import BaseModel
from starlette.websockets import WebSocketState

logger = logging.getLogger(__name__)

ROOM_TTL = 1800  # 30 minutes idle → drop
ROUNDS = 10      # iterated over 10 rounds; highest total after round 10 wins

MOVES = ("cooperate", "defect")

# (my move, opponent move) → (my points, opponent points)
PAYOFF = {
    ("cooperate", "cooperate"): (3, 3),
    ("defect", "defect"): (1, 1),
    ("cooperate", "defect"): (0, 5),
    ("defect", "cooperate"): (5, 0),
}


# ── In-memory state ──────────────────────────────────────────────────────────


@dataclass
class WebPlayer:
    name: str
    slot: str  # "A" (host) or "B"
    ws: WebSocket | None = None
    connected: bool = False
    move: str | None = None  # this round's pick — private until both are in
    total: int = 0


@dataclass
class PrisonerRoom:
    room_id: str
    code: str
    players: list[WebPlayer] = field(default_factory=list)  # [0]=A (host), [1]=B
    round: int = 1
    revealed: bool = False           # both picked & round resolved (moves shown)
    # history[i] = {"A": move, "B": move, "gain_A": int, "gain_B": int}
    history: list = field(default_factory=list)
    match_winner: str | None = None  # None | "A" | "B" | "tie"
    created_at: float = field(default_factory=time.time)
    last_activity: float = field(default_factory=time.time)


rooms: dict[str, PrisonerRoom] = {}   # room_id → room
codes: dict[str, str] = {}            # CODE → room_id


def _new_code() -> str:
    for _ in range(50):
        code = "".join(random.choices(string.ascii_uppercase, k=4))
        if code not in codes:
            return code
    # Astronomically unlikely fallback.
    return "".join(random.choices(string.ascii_uppercase, k=4))


# ── Request models ───────────────────────────────────────────────────────────


class JoinBody(BaseModel):
    name: str = "Player"


# ── API router ───────────────────────────────────────────────────────────────

router = APIRouter(prefix="/api/v1/prisoner", tags=["prisoner"])


@router.post("/rooms")
async def create_room(body: JoinBody):
    """Create a fresh room and register the caller as slot A (the host)."""
    name = (body.name or "Player").strip()[:32] or "Player"
    room_id = secrets.token_hex(4)
    code = _new_code()
    room = PrisonerRoom(room_id=room_id, code=code)
    room.players.append(WebPlayer(name=name, slot="A"))
    rooms[room_id] = room
    codes[code] = room_id
    logger.info("prisoner room created code=%s id=%s host=%s", code, room_id, name)
    return {"room_id": room_id, "code": code}


@router.post("/rooms/{code}/join")
async def join_room(code: str, body: JoinBody):
    """Join an existing room by its 4-letter code as slot B (the second player)."""
    code = (code or "").strip().upper()
    room_id = codes.get(code)
    room = rooms.get(room_id) if room_id else None
    if not room:
        raise HTTPException(404, "Room not found")
    name = (body.name or "Player").strip()[:32] or "Player"

    # Allow rejoin by same name (someone reloading the tab).
    for p in room.players:
        if p.name == name:
            return {"room_id": room.room_id}

    if len(room.players) >= 2:
        raise HTTPException(409, "Room full")

    slot = "A" if not room.players else "B"
    room.players.append(WebPlayer(name=name, slot=slot))
    room.last_activity = time.time()
    await _broadcast_state(room)
    return {"room_id": room.room_id}


# ── WebSocket handler ────────────────────────────────────────────────────────


async def prisoner_websocket(websocket: WebSocket, room_id: str):
    room = rooms.get(room_id)
    if not room:
        await websocket.close(code=4004, reason="Room not found")
        return

    await websocket.accept()
    player: WebPlayer | None = None

    try:
        while True:
            data = await websocket.receive_json()
            if not isinstance(data, dict):
                continue
            mtype = data.get("type")

            if mtype == "identify":
                name = str(data.get("name") or "").strip()[:32]
                player = _attach_player(room, name, websocket)
                if player is None:
                    await _send(websocket, {
                        "type": "error",
                        "message": "You're not in this room — join with the code first.",
                    })
                    continue
                room.last_activity = time.time()
                await _broadcast_state(room)

            elif mtype == "choice":
                if player is not None:
                    await _handle_choice(room, player, data)

            elif mtype == "rematch":
                if player is not None:
                    _reset_match(room)
                    room.last_activity = time.time()
                    await _broadcast_state(room)

            elif mtype == "ping":
                await _send(websocket, {"type": "pong"})

    except WebSocketDisconnect:
        pass
    except Exception:
        logger.exception("prisoner_websocket error in room %s", room_id)
    finally:
        if player is not None:
            player.connected = False
            player.ws = None
            try:
                await _broadcast_state(room)
            except Exception:
                pass


def _attach_player(room: PrisonerRoom, name: str, ws: WebSocket) -> WebPlayer | None:
    """Match an identify by name to a registered player and attach the socket.

    Returns None if the name isn't registered in this room (they must join via
    the REST endpoint first). Same-name identify reattaches — this is how rejoin
    after a disconnect/reload works.
    """
    if not name:
        return None
    for p in room.players:
        if p.name == name:
            p.ws = ws
            p.connected = True
            return p
    return None


async def _handle_choice(room: PrisonerRoom, player: WebPlayer, data: dict) -> None:
    if room.match_winner is not None:
        return  # match's over — wait for a rematch
    if len(room.players) < 2:
        await _send_error(player, "Waiting for an opponent to join.")
        return

    move = data.get("move")
    if move not in MOVES:
        return  # bad input — ignore

    # A choice arriving while last round is still revealed starts a fresh round.
    if room.revealed:
        _start_round(room)

    if player.move is not None:
        return  # already picked this round — ignore double-choices

    player.move = move

    if all(p.move is not None for p in room.players):
        _resolve_round(room)

    room.last_activity = time.time()
    await _broadcast_state(room)


def _start_round(room: PrisonerRoom) -> None:
    room.round += 1
    room.revealed = False
    for p in room.players:
        p.move = None


def _resolve_round(room: PrisonerRoom) -> None:
    a, b = room.players[0], room.players[1]
    gain_a, gain_b = PAYOFF[(a.move, b.move)]
    a.total += gain_a
    b.total += gain_b
    room.history.append({
        "A": a.move, "B": b.move, "gain_A": gain_a, "gain_B": gain_b,
    })
    room.revealed = True
    if room.round >= ROUNDS:
        if a.total > b.total:
            room.match_winner = a.slot
        elif b.total > a.total:
            room.match_winner = b.slot
        else:
            room.match_winner = "tie"


def _reset_match(room: PrisonerRoom) -> None:
    room.round = 1
    room.revealed = False
    room.match_winner = None
    room.history = []
    for p in room.players:
        p.move = None
        p.total = 0


# ── Broadcast helpers ────────────────────────────────────────────────────────


async def _send(ws: WebSocket | None, msg: dict) -> None:
    if ws and ws.client_state == WebSocketState.CONNECTED:
        try:
            await ws.send_json(msg)
        except Exception:
            pass


async def _send_error(player: WebPlayer, message: str) -> None:
    await _send(player.ws, {"type": "error", "message": message})


def _history_for(room: PrisonerRoom, me: WebPlayer, opp: WebPlayer | None) -> list:
    """Translate slot-keyed history into this viewer's me/opp perspective."""
    ms, os_ = me.slot, (opp.slot if opp else ("B" if me.slot == "A" else "A"))
    out = []
    for h in room.history:
        out.append({
            "me": h[ms],
            "opp": h[os_],
            "gain_me": h["gain_" + ms],
            "gain_opp": h["gain_" + os_],
        })
    return out


def _state_msg(room: PrisonerRoom, me: WebPlayer) -> dict:
    opp = next((p for p in room.players if p.slot != me.slot), None)
    match_winner = None
    if room.match_winner is not None:
        match_winner = (
            "tie" if room.match_winner == "tie"
            else ("you" if room.match_winner == me.slot else "opp")
        )
    hist = _history_for(room, me, opp)
    last = hist[-1] if hist else {"me": None, "opp": None}
    return {
        "type": "state",
        "code": room.code,
        "round": room.round,
        "rounds": ROUNDS,
        "revealed": room.revealed,
        "totals": {"me": me.total, "opp": (opp.total if opp else 0)},
        "your_move": me.move,
        # Opponent's move stays hidden until both have picked & the round resolves.
        "opp_move": (opp.move if (opp and room.revealed) else None),
        "opp_chosen": bool(opp and opp.move is not None),
        "last": {"me": last["me"], "opp": last["opp"]},
        "history": hist,
        "match_winner": match_winner,
        "players": [
            {"name": p.name, "slot": p.slot, "total": p.total, "connected": p.connected}
            for p in room.players
        ],
        "you": me.slot,
    }


async def _broadcast_state(room: PrisonerRoom) -> None:
    for p in room.players:
        await _send(p.ws, _state_msg(room, me=p))


# ── Stale room cleanup ───────────────────────────────────────────────────────


async def cleanup_stale_prisoner_rooms() -> None:
    """Background task: drop rooms that have sat idle past the TTL."""
    while True:
        await asyncio.sleep(60)
        try:
            now = time.time()
            stale = [
                rid for rid, room in rooms.items()
                if (now - room.last_activity) > ROOM_TTL
            ]
            for rid in stale:
                room = rooms.pop(rid, None)
                if room:
                    codes.pop(room.code, None)
                    for p in room.players:
                        await _send(p.ws, {
                            "type": "error",
                            "message": "Room expired after inactivity.",
                        })
        except Exception:
            logger.exception("cleanup_stale_prisoner_rooms error — loop continues")

File: web/static/prisoner.js
// SharpLab HQ — Iterated Prisoner's Dilemma. Free 2-player online game over WebSockets.
// Web-native multiplayer: rooms are created straight from the browser (no Discord
// token), players share a 4-letter code, and simultaneous secret moves flow over a
// socket to /ws/prisoner/{room_id}. Iterated over 10 rounds — highest total wins.
// No coins involved.

const app = document.getElementById("app");
const navRight = document.getElementById("navRight");

// ── Helpers (copied verbatim from rps.js / casino.js) ──
const num = (n) => (n == null ? "—" : Number(n).toLocaleString());
const esc = (s) =>
  String(s == null ? "" : s).replace(/[&<>"']/g, (c) =>
    ({ "&": "&amp;", "<": "&lt;", ">": "&gt;", '"': "&quot;", "'": "&#39;" }[c]));

async function getJSON(url) {
  const r = await fetch(url, { credentials: "include" });
  if (!r.ok) return { _status: r.status };
  return r.json();
}

// ── Nav (login / logout) — mirrors casino.js (reads state.balance) ──
function renderNav(user) {
  if (!user) {
    navRight.innerHTML = `<a class="btn" href="/api/v1/auth/discord/login">Sign in with Discord</a>`;
    return;
  }
  const av = user.avatar ? `https://cdn.discordapp.com/avatars/${user.id}/${user.avatar}.png` : null;
  navRight.innerHTML = `<div class="userbar">
    ${av ? `<img class="avatar" src="${av}" alt="">` : `<div class="avatar"></div>`}
    <span>${esc(user.username)}</span>
    <span class="coinschip" title="Casino coins">🪙 ${num(state.balance)}</span>
    <a class="btn ghost" href="/api/v1/auth/logout">Sign out</a></div>`;
}

// ── State ──
const state = { me: null, balance: 0 };

// ── Toast (copied verbatim from casino.js) ──
function toast(msg) {
  const t = document.createElement("div");
  t.className = "cardtoast";
  t.textContent = msg;
  document.body.appendChild(t);
  setTimeout(() => t.remove(), 3200);
}

// ── Coins hub (copied verbatim from casino.js) ──
const EARN_WAYS = [
  ["💬", "Chat in the server", "5 coins per message, up to 500 a day"],
  ["🎯", "Log a bet", "50 coins for logging a sports bet with /bet log"],
  ["📈", "Log a trade", "50 coins for recording a stock or option trade"],
  ["🏀", "Daily pick'em", "25 coins per pick — plus a coin payout when your pick wins"],
  ["🎁", "Free daily pack", "Open one free card pack every day — pure upside"],
  ["🃏", "Complete a set", "One-time coin bonus for owning every card in a set"],
  ["♻️", "Quick-sell dupes", "Sell duplicate cards back for coins in Discord"],
  ["🎮", "Win in the casino", "Win at /casino or /play — playing no longer hands out free coins"],
];
function showCoinsHub() {
  if (document.querySelector(".hubov")) return;
  const rows = EARN_WAYS.map(
    ([i, t, d]) => `<div class="hubrow"><div class="hubicon">${i}</div>
      <div><div class="hubt">${esc(t)}</div><div class="hubd">${esc(d)}</div></div></div>`
  ).join("");
  const ov = document.createElement("div");
  ov.className = "hubov";
  ov.innerHTML = `<div class="hubcard">
    <div class="hubhead"><h3>Ways to earn 🪙</h3><button class="hubx" aria-label="Close">✕</button></div>
    ${rows}
    <div class="hubfoot">Your balance: <b>🪙 ${num(state.balance)}</b></div>
  </div>`;
  ov.addEventListener("click", (e) => { if (e.target === ov || e.target.closest(".hubx")) ov.remove(); });
  document.body.appendChild(ov);
}
document.addEventListener("click", (e) => { if (e.target.closest(".coinschip")) showCoinsHub(); });

const REDUCE = matchMedia("(prefers-reduced-motion: reduce)").matches;

// ── Moves ──
const EMOJI = { cooperate: "🤝", defect: "🔪" };
const LABEL = { cooperate: "Cooperate", defect: "Defect" };
const MOVES = ["cooperate", "defect"];

// ── Game state ──
const $ = (id) => document.getElementById(id);
const game = {
  ws: null,
  roomId: null,
  code: null,
  name: "",
  you: null,        // "A" | "B"
  totals: { me: 0, opp: 0 },
  round: 1,
  rounds: 10,
  revealed: false,
  yourMove: null,
  oppMove: null,
  oppChosen: false,
  last: { me: null, opp: null },
  history: [],
  matchWinner: null, // null | "you" | "opp" | "tie"
  players: [],
};

// ── REST: create / join a room ──
async function postJSON(url, body) {
  const r = await fetch(url, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body || {}),
  });
  const j = await r.json().catch(() => ({}));
  if (!r.ok) j._status = r.status;
  return j;
}

async function createRoom() {
  const name = readName();
  if (!name) return;
  const res = await postJSON("/api/v1/prisoner/rooms", { name });
  if (res._status || !res.room_id) return toast("❌ Couldn't create a room.");
  game.name = name;
  game.code = res.code;
  connect(res.room_id);
}

async function joinRoom() {
  const name = readName();
  if (!name) return;
  const code = ((game._codeInput && game._codeInput.value) || "").trim().toUpperCase();
  if (code.length !== 4) return toast("Enter the 4-letter room code.");
  const res = await postJSON(`/api/v1/prisoner/rooms/${code}/join`, { name });
  if (res._status === 404) return toast("❌ No room with that code.");
  if (res._status === 409) return toast("❌ That room is full.");
  if (res._status || !res.room_id) return toast("❌ Couldn't join the room.");
  game.name = name;
  game.code = code;
  connect(res.room_id);
}

function readName() {
  const el = $("pdName");
  const name = ((el && el.value) || "").trim();
  if (!name) { if (el) el.focus(); toast("Enter a display name first."); return null; }
  return name.slice(0, 32);
}

// ── WebSocket ──
function connect(roomId) {
  game.roomId = roomId;
  try {
    const proto = location.protocol === "https:" ? "wss" : "ws";
    const ws = new WebSocket(`${proto}://${location.host}/ws/prisoner/${roomId}`);
    game.ws = ws;
    ws.onopen = () => ws.send(JSON.stringify({ type: "identify", name: game.name }));
    ws.onmessage = (ev) => {
      let msg;
      try { msg = JSON.parse(ev.data); } catch { return; }
      handleMessage(msg);
    };
    ws.onerror = () => toast("⚠️ Connection error.");
    ws.onclose = () => {
      if (game.ws === ws) {
        game.ws = null;
        const s = $("pdStatus");
        if (s) { s.textContent = "Disconnected."; s.className = "pd-status wait"; }
      }
    };
  } catch {
    toast("⚠️ Could not open a connection.");
  }
}

function handleMessage(msg) {
  if (!msg || typeof msg !== "object") return;
  if (msg.type === "error") return toast("⚠️ " + (msg.message || "error"));
  if (msg.type === "pong") return;
  if (msg.type !== "state") return;

  game.code = msg.code || game.code;
  game.round = msg.round || 1;
  game.rounds = msg.rounds || 10;
  game.revealed = !!msg.revealed;
  game.totals = msg.totals || { me: 0, opp: 0 };
  game.yourMove = msg.your_move || null;
  game.oppMove = msg.opp_move || null;
  game.oppChosen = !!msg.opp_chosen;
  game.last = msg.last || { me: null, opp: null };
  game.history = Array.isArray(msg.history) ? msg.history : [];
  game.matchWinner = msg.match_winner || null;
  game.players = msg.players || [];
  game.you = msg.you;

  renderGame();
}

function sendChoice(move) {
  if (!game.ws || game.ws.readyState !== WebSocket.OPEN) return;
  if (game.matchWinner) return;
  if (game.players.length < 2) return toast("Waiting for an opponent to join.");
  // Can pick when starting a fresh round, or when the last round is revealed
  // (which begins the next round). Block while waiting on the opponent.
  if (game.yourMove != null && !game.revealed) return;
  game.ws.send(JSON.stringify({ type: "choice", move }));
}

function sendRematch() {
  if (!game.ws || game.ws.readyState !== WebSocket.OPEN) return;
  game.ws.send(JSON.stringify({ type: "rematch" }));
}

// ── Rendering ──
function buildLobby() {
  app.innerHTML = `
    <div class="pd-head">
      <h1>🤝 Prisoner's Dilemma 🔪</h1>
      <p>Free 2-player online game — 10 rounds. Create a room, share the code, and play.</p>
    </div>
    <div class="pd-wrap">
      <div class="card pd-lobby">
        <div class="pd-rules">
          Each round you both <b>secretly</b> choose to <b>Cooperate</b> or <b>Defect</b>.
          Points score by the classic matrix; after 10 rounds the higher total wins.
          <div class="pd-payoff">
            <div class="pd-payoff-cell">Both cooperate<b>+3 / +3</b></div>
            <div class="pd-payoff-cell">Both defect<b>+1 / +1</b></div>
            <div class="pd-payoff-cell">You defect, they cooperate<b>+5 / +0</b></div>
            <div class="pd-payoff-cell">You cooperate, they defect<b>+0 / +5</b></div>
          </div>
        </div>
        <div class="pd-field">
          <label for="pdName">Display name</label>
          <input class="pd-input" id="pdName" type="text" maxlength="32" autocomplete="off"
                 spellcheck="false" placeholder="Your name" value="${esc(defaultName())}" />
        </div>
        <div class="pd-actions">
          <button class="btn primary big" id="pdCreate">Create room</button>
        </div>
        <div class="pd-sep">— or join a room —</div>
        <div class="pd-field">
          <label for="pdCode">Room code</label>
          <div class="pd-joinrow">
            <input class="pd-input code" id="pdCode" type="text" maxlength="4" autocomplete="off"
                   spellcheck="false" placeholder="ABCD" />
            <button class="btn" id="pdJoin">Join</button>
          </div>
        </div>
      </div>
    </div>`;
  game._codeInput = $("pdCode");
  $("pdCreate").addEventListener("click", createRoom);
  $("pdJoin").addEventListener("click", joinRoom);
  game._codeInput.addEventListener("keydown", (e) => { if (e.key === "Enter") joinRoom(); });
  game._codeInput.addEventListener("input", (e) => {
    e.target.value = e.target.value.toUpperCase().replace(/[^A-Z]/g, "").slice(0, 4);
  });
}

function defaultName() {
  if (state.me && state.me.user && state.me.user.username) return state.me.user.username;
  return "";
}

function renderGame() {
  const twoPlayers = game.players.length >= 2;
  const me = game.players.find((p) => p.slot === game.you);
  const opp = game.players.find((p) => p.slot !== game.you);

  const showReveal = game.revealed;
  const waiting = game.yourMove != null && !game.revealed; // picked, awaiting opp

  // Status line.
  let statusText = "", statusClass = "pd-status wait";
  if (!twoPlayers) {
    statusText = "Waiting for opponent…";
  } else if (game.matchWinner === "you") {
    statusText = "You win the match! 🎉"; statusClass = "pd-status win";
  } else if (game.matchWinner === "opp") {
    statusText = "You lost the match."; statusClass = "pd-status lose";
  } else if (game.matchWinner === "tie") {
    statusText = "Dead tie — nobody wins."; statusClass = "pd-status draw";
  } else if (showReveal) {
    const gm = game.last.me, go = game.last.opp;
    if (gm && go) {
      const verb = (m) => (m === "cooperate" ? "cooperated" : "defected");
      statusText = `You ${verb(gm)}, they ${verb(go)}.`;
      statusClass = "pd-status yours";
    } else {
      statusText = "Round scored — choose again.";
    }
  } else if (waiting) {
    statusText = "Waiting for opponent…"; statusClass = "pd-status wait";
  } else {
    statusText = "Make your choice"; statusClass = "pd-status yours";
  }

  // Faces: your side + opponent side.
  const yourMove = (showReveal || waiting) ? game.yourMove : null;
  const yourFace = faceHtml(yourMove, showReveal || waiting, false);
  const oppReveal = showReveal ? game.oppMove : null;
  const oppFace = faceHtml(oppReveal, showReveal, !showReveal && game.oppChosen);

  const lastGainMe = (showReveal && game.history.length) ? game.history[game.history.length - 1].gain_me : null;
  const lastGainOpp = (showReveal && game.history.length) ? game.history[game.history.length - 1].gain_opp : null;

  const totMe = me ? me.total : game.totals.me;
  const totOpp = opp ? opp.total : game.totals.opp;

  const pchip = (p, isMe) => {
    if (!p) return `<div class="pd-pchip">
      <span class="pd-pname muted">waiting…</span>
      <span class="pd-score">0</span>
    </div>`;
    const active = !game.matchWinner && twoPlayers;
    const lead = twoPlayers && p.total > (isMe ? totOpp : totMe);
    return `<div class="pd-pchip${active ? " active" : ""}${lead ? " lead" : ""}">
      <span class="pd-pname">${esc(p.name)}</span>
      ${isMe ? `<span class="pd-pyou">(you)</span>` : ""}
      <span class="pd-dot${p.connected ? " on" : ""}" title="${p.connected ? "connected" : "offline"}"></span>
      <span class="pd-score">${p.total}</span>
    </div>`;
  };

  const buttonsLocked = !twoPlayers || game.matchWinner || waiting;
  const buttons = MOVES.map((m) => {
    const chosen = game.yourMove === m && (waiting || showReveal);
    return `<button class="pd-choice ${m}${chosen ? " chosen" : ""}" data-move="${m}"
      ${buttonsLocked ? "disabled" : ""} title="${LABEL[m]}">
        <span class="pd-choice-emoji">${EMOJI[m]}</span>
        <span class="pd-choice-label">${LABEL[m]}</span>
      </button>`;
  }).join("");

  const banner = game.matchWinner
    ? `<div class="pd-banner"><button class="btn primary" id="pdRematch">Rematch</button></div>`
    : "";

  app.innerHTML = `
    <div class="pd-head">
      <h1>🤝 Prisoner's Dilemma 🔪</h1>
    </div>
    <div class="pd-wrap">
      <div class="card" style="display:flex;flex-direction:column;gap:var(--s-4)">
        <div class="pd-codebox">
          <div class="pd-codelabel">Room code</div>
          <div class="pd-code" id="pdCodeBox" title="Click to copy">${esc(game.code || "····")}</div>
          <span class="pd-copyhint">Share this code with a friend to play. 10 rounds.</span>
        </div>
        <div class="pd-round">Round <b>${Math.min(game.round, game.rounds)}</b> / ${game.rounds}</div>
        <div class="pd-players">${pchip(me, true)}<span class="pd-vs">vs</span>${pchip(opp, false)}</div>
        <div class="pd-arena">
          <div class="pd-side">
            <div class="pd-face${showReveal ? " reveal " + (yourMove || "") : ""}">${yourFace}</div>
            <div class="pd-side-label">You</div>
            <div class="pd-gain ${lastGainMe ? "pos" : "zero"}">${lastGainMe != null ? "+" + lastGainMe : ""}</div>
          </div>
          <div class="pd-arena-vs">${showReveal ? "⚖️" : "⋯"}</div>
          <div class="pd-side">
            <div class="pd-face${showReveal ? " reveal " + (oppReveal || "") : ""}">${oppFace}</div>
            <div class="pd-side-label">${esc(opp ? opp.name : "Opponent")}</div>
            <div class="pd-gain ${lastGainOpp ? "pos" : "zero"}">${lastGainOpp != null ? "+" + lastGainOpp : ""}</div>
          </div>
        </div>
        <div class="${statusClass}" id="pdStatus">${statusText}</div>
        <div class="pd-choices">${buttons}</div>
        ${historyStrip()}
        ${banner}
      </div>
    </div>`;

  const choicesEl = app.querySelector(".pd-choices");
  choicesEl.addEventListener("click", (e) => {
    const btn = e.target.closest(".pd-choice");
    if (!btn || btn.disabled) return;
    sendChoice(btn.dataset.move);
  });
  const codeBox = $("pdCodeBox");
  if (codeBox) codeBox.addEventListener("click", copyCode);
  const rematch = $("pdRematch");
  if (rematch) rematch.addEventListener("click", sendRematch);
}

// A face: shows the move emoji when `reveal` is true, a hidden fist when the
// player has picked but it's still secret, and a placeholder otherwise.
function faceHtml(move, reveal, chosenHidden) {
  if (reveal && move) return `<span class="pd-face-in">${EMOJI[move]}</span>`;
  if (chosenHidden || (reveal && !move)) return `<span class="pd-face-hidden">✊</span>`;
  return `<span class="pd-face-empty">·</span>`;
}

function historyStrip() {
  if (!game.history.length) return "";
  const cells = game.history.map((h, i) => `<div class="pd-hcell" title="R${i + 1}">
      <span class="pd-hn">${i + 1}</span>
      <span class="pd-hmoves">${EMOJI[h.me] || "·"}${EMOJI[h.opp] || "·"}</span>
      <span class="pd-hpts">${h.gain_me}/${h.gain_opp}</span>
    </div>`).join("");
  return `<div class="pd-history">
    <div class="pd-history-head">History — you / opp</div>
    <div class="pd-hstrip">${cells}</div>
  </div>`;
}

function copyCode() {
  if (!game.code) return;
  if (navigator.clipboard && navigator.clipboard.writeText) {
    navigator.clipboard.writeText(game.code).then(
      () => toast("📋 Code copied: " + game.code),
      () => toast("Code: " + game.code)
    );
  } else {
    toast("Code: " + game.code);
  }
}

async function main() {
  const me = await getJSON("/api/v1/hq/me");
  const loggedIn = me && me.authenticated;
  state.me = loggedIn ? me : null;
  state.balance = loggedIn ? (me.balance || 0) : 0;
  renderNav(loggedIn ? me.user : null);
  buildLobby();
}

main();
